src/hp/magicsock/conn.rs

```diff
@@ -179,525 +189,267 @@ pub struct Inner {
-        if let Some(conn6) = pconn6.clone() {
-            net_checker.get_stun_conn6 = Some(Arc::new(Box::new(move || conn6.as_socket())));
+        if self.handle_disco_message(dm.buf.clone(), ipp, Some(dm.src.clone())) {
+            // Message was internal, do not bubble up.
+            return true;
         }
 
-        let c = Conn(Arc::new(Inner {
-            name,
-            on_endpoints,
-            on_derp_active,
-            idle_for,
-            on_net_info,
-            port: AtomicU16::new(port),
-            port_mapper,
-            net_checker,
-            public_key: private_key.public_key().into(),
-            last_net_check_report: Default::default(),
-            no_v4_send: AtomicBool::new(false),
-            pconn4,
-            pconn6,
-            socket_endpoint4: SocketEndpointCache::default(),
-            socket_endpoint6: SocketEndpointCache::default(),
-            on_stun_receive: Default::default(),
-            state: ConnState::new(private_key).into(),
-            closing: AtomicBool::new(false),
-            derp_recv_ch,
-            derp_map: Default::default(),
-            peer_map: Default::default(),
-        }));
+        let ep_fake_wg_addr = {
+            let peer_map = tokio::task::block_in_place(|| self.peer_map.blocking_read());
+            peer_map
+                .endpoint_for_node_key(&dm.src)
+                .map(|ep| ep.fake_wg_addr)
+        };
 
-        Ok(c)
-    }
+        match ep_fake_wg_addr {
+            Some(ep_fake_wg_addr) => {
+                b[..dm.buf.len()].copy_from_slice(&dm.buf);
 
-    /// Sets a STUN packet processing func that does nothing.
-    async fn ignore_stun_packets(&self) {
-        *self.on_stun_receive.write().await = None;
-    }
+                // Update RecvMeta structure accordingly.
+                meta.len = dm.buf.len();
+                meta.stride = dm.buf.len();
+                meta.addr = ep_fake_wg_addr;
 
-    pub(super) fn is_closing(&self) -> bool {
-        self.closing.load(Ordering::Relaxed)
+                // if stats := c.stats.Load(); stats != nil {
+                // 	stats.UpdateRxPhysical(ep.nodeAddr, ipp, dm.n)
+                // }
+                false
+            }
+            None => {
+                // We don't know anything about this node key, nothing to record or process.
+                true
+            }
+        }
     }
 
-    /// Called (in a new task) by `periodic_re_stun_timer` when periodic STUNs are active.
-    async fn do_periodic_stun(&self) {
-        self.re_stun("periodic").await;
+    fn handle_disco_message(
+        &self,
+        msg: Vec<u8>,
+        src: SocketAddr,
+        derp_node_src: Option<key::node::PublicKey>,
+    ) -> bool {
+        let (s, r) = flume::bounded(1);
+        self.derp_sender.send(DerpMessage::HandleDiscoMessage {
+            msg,
+            src,
+            derp_node_src,
+            s,
+        });
+        r.recv().expect("dropped sender")
```

Comment:
this `r.recv()` blocks the executor. The `DerpHandler::run` loop is unable to receive the `DerpMessage::HandleDiscoMessage`, and we never process the `msg`, and never send anything back on `s`. This doesn't fix everything, but we are able to connect and pass data between `m1` & `m2` when this is fixed

Reply:
fixed